### PR TITLE
core: allow pausing and un-pausing of leader broker routine

### PR DESCRIPTION
**Pull request by jrasell**
Test failure: percy-ui and unrelated
Related to https://github.com/hashicorp/nomad/issues/11638
Commits: split to make review easier

This changes allows operators to control whether or not the eval broker is running on the leader server. This is useful in outage situations where administrators wish to stop any work being available to the scheduler workers. It is also requisite work for future work which will allow deletion of evaluations which have not been processed.

Alongside pausing the eval broker, the blocked evals process is also disabled at the same time. This sub-process pushes evaluations into the eval broker and is also restored from the state store using the same process as the eval broker. It therefore makes sense, considering the end goal, to disable both processes for state consistency.

To ensure operators pausing/un-pausing the broker doesn't conflict with leadership transitions a mutex is used to control access to the eval broker and blocked evals processes. This is used along with a leadership check when changing the broker status which requires taking into account operator configuration.

The issue mentions disabling the `reapFailedEvaluations` process, however, I don't believe this is required for this change as disabling the eval broker also flushes all stored evaluation data. It therefore seemed safer to leave this alone and not require additional coordination.

The new `operator scheduler` commands allow inspecting and modifying the scheduler config. This is useful as it doesn't require you to supply the full payload object or remember the exact curl command to run.

### CODE REVIEW

File: command/operator_scheduler_set_config.go
Comment (schmichael) [inline command/operator_scheduler_set_config.go:83]:
```suggestion
		o.Ui.Error(fmt.Sprintf("Error querying for scheduler configuration: %s", err))
```

File: nomad/eval_endpoint.go
Comment (schmichael) [inline nomad/eval_endpoint.go:134]: Will this cause the workers to try to Dequeue in a tight loop that does nothing? Logging errors is good to avoid, but we also have 0 visibility into when this is hit.

When there are no evals to be processed the eval broker blocks the Dequeue RPC until there is an eval to be processed. Perhaps we need a similar mechanism here so that instead of returning when the eval broker is disabled, we block until it is enabled or there is a leadership transition?

Comment (jrasell) [reply nomad/eval_endpoint.go:134]: I didn't realise quite how tight that loop was; thanks for pointing this out.

Yeah that sort of functionality makes sense and I will look into how this could be implemented over the coming days.

**Comment by jrasell** [comment]
I am also wondering if we need evalBroker enabled checks on other RPC handlers such as `Ack` and `Nack`?

File: command/operator_scheduler_set_config.go
Comment (DerekStrickland) [inline command/operator_scheduler_set_config.go:153]: Out of curiosity, what happens if they set more than one of these to true?

**Review by DerekStrickland — APPROVED** [review]
Great job on a difficult task! I left some non-blocking comments and questions. I'd personally wait on schmichael's review to merge, but you may have already been through that and it's just not updated.

File: e2e/operator_scheduler/input/basic.nomad
Comment (tgross) [inline e2e/operator_scheduler/input/basic.nomad:17]: It might be worth dropping this down to 1s and making it a batch job so that we can guarantee a quick exit once we've turned the broker back on; that'll reduce the risk of test flakiness on the `WaitForJobStop` (which definitely _should_ be enough time, but...)

**Review by lgfa29 — APPROVED** [review]
Nice work!

Lots of minor comments, but one thing that may be worth looking into is to disable deployment monitoring for the `nomad job run` command if the eval broker is disabled, otherwise it will get stuck:

```console
$ nomad run example.nomad
==> 2022-06-10T11:11:17-04:00: Monitoring evaluation "076115a0"
    2022-06-10T11:11:17-04:00: Evaluation triggered by job "countdash"
==> 2022-06-10T11:11:18-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:19-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:20-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:21-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:22-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:23-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:24-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:25-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:26-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:27-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:28-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:29-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:30-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:31-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:32-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:33-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:34-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:35-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:36-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:37-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:38-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:39-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:40-04:00: Monitoring evaluation "076115a0"
==> 2022-06-10T11:11:41-04:00: Monitoring evaluation "076115a0"
```

File: helper/broker/notify.go
Comment (tgross) [inline helper/broker/notify.go:93]: I spent a half hour writing up a whole thing about how I think we can potentially block here and cause updates to get dropped but then re-read and realized that it's not an issue because this function just returns and isn't responsible for new updates. So I'm leaving this comment here to spare another reviewer the same effort. 😊

File: website/content/docs/commands/operator/scheduler-set-config.mdx
Comment (tgross) [inline website/content/docs/commands/operator/scheduler-set-config.mdx:32]:
```suggestion
  limit, if the client has excess memory capacity. Tasks must specify `memory_max`
```

(maybe link it too?)

File: website/content/docs/commands/operator/scheduler-set-config.mdx
Comment (tgross) [inline website/content/docs/commands/operator/scheduler-set-config.mdx:35]:
```suggestion
- `-reject-job-registration` - When true, the server will return permission denied
```

File: nomad/eval_endpoint.go
Comment (jrasell) [reply nomad/eval_endpoint.go:148]: The server workers are the only processes which call this endpoint and will only utilise the meta if the [response eval object is not nil](https://github.com/hashicorp/nomad/blob/main/nomad/worker.go#L484-L487). Interestingly, the worker uses the meta that is set [here](https://github.com/hashicorp/nomad/blob/main/nomad/eval_endpoint.go#L157-L159) and noting which is set with the `setQueryMeta` function call. Therefore skipping the `setQueryMeta` call is fine here.

File: command/operator_scheduler_set_config.go
Comment (DerekStrickland) [reply command/operator_scheduler_set_config.go:155]: That answers it.

File: command/operator_scheduler_set_config.go
Comment (jrasell) [reply command/operator_scheduler_set_config.go]: I went with `preempt-*-scheduler` to keep it inline somewhat with the API object. I also changes the prediction like you pointed out.

**Comment by jrasell** [comment]
> Nice work!
> 
> Lots of minor comments, but one thing that may be worth looking into is to disable deployment monitoring for the `nomad job run` command if the eval broker is disabled, otherwise it will get stuck:
> 
> ```
> $ nomad run example.nomad
> ==> 2022-06-10T11:11:17-04:00: Monitoring evaluation "076115a0"
>     2022-06-10T11:11:17-04:00: Evaluation triggered by job "countdash"
> ==> 2022-06-10T11:11:18-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:19-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:20-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:21-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:22-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:23-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:24-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:25-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:26-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:27-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:28-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:29-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:30-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:31-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:32-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:33-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:34-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:35-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:36-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:37-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:38-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:39-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:40-04:00: Monitoring evaluation "076115a0"
> ==> 2022-06-10T11:11:41-04:00: Monitoring evaluation "076115a0"
> ```
> 
> Also missing a `CHANGELOG` entry.

Changelog has been added.

I quite like the UX of the job submission in situation when the eval broker is paused. It shows the evaluation is making no progress which is exactly correct and will unblock if the eval broker is enabled and the eval being watched changes state. In order to avoid this we would need to make an extra call to the API for another that monitors evals such as deployments promotion, job stop, and others. That commands also have the detach flag which can be used to avoid this.

**Comment by tgross** [comment]
That the "Monitoring evaluation" message falls inside the loop instead of at the top feels almost accidental; in most cases we'd only ever see it once. If we move it up so that it's outside the loop we still the effect of "hey we're waiting here" without making a bunch of noise in the terminal?

**Comment by lgfa29** [comment]
> That the "Monitoring evaluation" message falls inside the loop instead of at the top feels almost accidental; in most cases we'd only ever see it once. If we move it up so that it's outside the loop we still the effect of "hey we're waiting here" without making a bunch of noise in the terminal?

Good point, I think that would work as well 👍

File: command/operator_scheduler_get_config.go
Comment (schmichael) [inline command/operator_scheduler_get_config.go:85]: Can we include `"Modify Index|%v", resp.ModifyIndex` to encourage use of CAS with these operations? Really don't want operators overwriting one another's configs unintentionally.

Comment (schmichael) [reply command/operator_scheduler_get_config.go:85]:
```suggestion
		fmt.Sprintf("Preemption SysBatch Scheduler|%v", schedConfig.PreemptionConfig.SysBatchSchedulerEnabled),
		fmt.Sprintf("Modify Index|%v", resp.ModifyIndex),
```

File: command/operator_scheduler_set_config.go
Comment (schmichael) [inline command/operator_scheduler_set_config.go:68]: Can we add `check-index` like `nomad job run` to allow and even encourage use of the existing CAS operator?

File: command/operator_scheduler_set_config.go
Comment (schmichael) [inline command/operator_scheduler_set_config.go:83]: If we add `-check-index` we should fail here if `resp.ModifyIndex != check-index` because that means someone else updated the scheduler config and the user could end up with unintentional results. Imagine during an outage:

1. Operator Alice runs `nomad operator scheduler set-config -reject-job-registration=true`
2. Operator Bob runs `nomad operator scheduler set-config -pause-eval-broker=true`

Without CAS there's no way to know which of the following will be the final state of the scheduler config:

1. RejectJobRegistration=true; PauseEvalBroker=false
2. RejectJobRegistration=false; PauseEvalBroker=true
3. RejectJobRegistration=true; PauseEvalBroker=true

Because the HTTP GETs and PUTs may interleave in any order.

Both Alice and Bob must _manually_ do a Get-before-Set and put the ModifyIndex form the result into the `-check-index` flag to ensure the resulting state is what they intend.

File: command/operator_scheduler_set_config.go
Comment (schmichael) [inline command/operator_scheduler_set_config.go:103]: We should always include a CAS index here:

1. Prefer the user's from `-check-index` if it is set (> 0)
2. If it is not set use `resp.ModifyIndex` to ensure this set does not interleave with another user's set in an unintended way.

This is not a concern for most of our other APIs like job registration because they're doing _atomic_ sets and will therefore either successfully set the object to exactly what they intend or receive an error. The flags on this command introduce PATCH-like semantics so we have to ensure we're not interleaving updates.

**Review by schmichael — APPROVED** [review]

**Pull request merged**
